docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+
+ - For singleton types, there are two possible answers:
+   - Each type requires one state in the discriminator but no bits to encode the value.
+   - Each type requires one pointer's worth of state to encode the value but no bits in the discriminator.
+   We choose the former, to maximise our ability to use the CPU's branch-on-zero instructions.
+ - Each integer requires one state in the discriminator and as many bits as its width defines for the payload.
+   For example, in the union `U8 | U16`, each requires one state in the discriminator (giving a 1-bit discriminator to represent both state) and 8 and 16 bits of payload, respectively.
+ - Floating point types follow the same rules as integers, with special handling if a floating point type is the type with the largest payload in a union type, described below.
+ - Pointers to object types (including external reference types) require as many bits as are required for an unambiguous pointer; however, all pointer types require only a single discriminator state because all pointees can be discriminated by inspecting their descriptor.
+   The number of bits required for a pointer varies depending on the pointer representation:
+   - 32-bit integer pointers require 29 bits of state.
+   - 64-bit integer pointers require either 45 or 52 bits, depending on the virtual address space size of the target.
+   - CHERI capabilities require 125 bits of state but provide a 'free' discriminator bit because the tag bit can be used to discriminate CHERI capabilities from non-pointer data.
+ - Value types are still being defined; however, it seems plausible that they will each require one descriptor state and as many payload bits as it has non-padding bits.
+   Making efficient use of padding bits as discriminators seems very difficult and so is likely to be left to a later ABI, if ever implemented.
+
+Once the payload sizes are all known, the algorithm for implementing the packing depends on the type of the largest payload.
+There are two special cases:
+
+If the largest type is a floating-point type and the payload of all other types plus the bits required for the discriminator will fit in the unused bits of the significand of a signalling NaN minus one then the top bit of the significand is reserved to discriminate between a real signalling NaN and another value and the remainder of the types are encoded in the spare bits, starting this algorithm again.
+If the remaining bits of the significand are not sufficient to encode the remaining states then this rule is ignored.
+
+If the largest type is a CHERI capability then the tag bit is used to differentiate pointer and non-pointer types.
+This leaves 128 bits to use for the remainder of the values.
+
+The size of the final object is the number of payload bits, plus the base-two logarithm of the number of discriminator states rounded up.
+The alignment of the result is the alignment of the most strongly aligned type.
```

Comment:
This paragraph seems premature given the next, or perhaps the next is intended as an example?